File: focus_todo_mobile/lib/sync/supabase_sync_transport.dart
import '../shared/models/task.dart';
import '../shared/models/task_sync_metadata.dart';
import 'canonical_task_mapper.dart';
import 'supabase_sync_service.dart';
import 'sync_transport.dart';
import 'supabase_config.dart';

class SupabaseTransportNotConfiguredError implements Exception {
  const SupabaseTransportNotConfiguredError();

  @override
  String toString() => 'Supabase sync is not configured';
}

class SupabaseSyncTransport implements SyncTransport {
  SupabaseSyncTransport({SupabaseConfig? config, SupabaseSyncService? service})
    : _config = config ?? getSupabaseConfig(),
      _service = service ?? const SupabaseSyncService();

  final SupabaseConfig _config;
  final SupabaseSyncService _service;

  bool isConfigured() => hasSupabaseConfig(_config);

  Future<void> _ensureInitialized() async {
    if (!isConfigured()) {
      throw const SupabaseTransportNotConfiguredError();
    }

    await initializeSupabaseIfConfigured(_config);
  }

  @override
  Future<void> pushTasks(List<CanonicalTask> outboundBatch) async {
    await _ensureInitialized();
    throw UnsupportedError('Bulk upsert is no longer used for Supabase sync');
  }

  @override
  Future<SyncWriteResult> writeTask(
    CanonicalTask task,
    String? expectedUpdatedAt,
  ) async {
    await _ensureInitialized();
    return _service.writeTodo(task, expectedUpdatedAt);
  }

  @override
  Future<List<CanonicalTask>> pullTasks() async {
    await _ensureInitialized();
    return _service.pullTodos();
  }

  @override
  Future<List<CanonicalTask>> syncOnce(
    List<Task> currentTasks,
    List<TaskSyncMetadata> currentSyncMetadata,
  ) async {
    throw UnimplementedError(
      'SupabaseSyncTransport does not support legacy syncOnce; use writeTask + pullTasks',
    );
  }
}

File: focus_todo_computer/focus-todo/src/sync/supabaseSyncTransport.ts
import {
  getSupabaseConfig,
  hasSupabaseConfig,
  type SupabaseConfig,
} from './supabaseConfig'
import { pullSupabaseTodos, writeSupabaseTodo } from './supabaseTodoSync'
import type { CanonicalTask } from './canonicalTaskMapper'
import type { SyncTransport, SyncWriteResult } from './syncTransport'

export class SupabaseTransportNotConfiguredError extends Error {
  constructor() {
    super('Supabase sync is not configured')
    this.name = 'SupabaseTransportNotConfiguredError'
  }
}

export class SupabaseSyncTransport implements SyncTransport {
  private readonly configured: boolean

  constructor(config: SupabaseConfig = getSupabaseConfig()) {
    this.configured = hasSupabaseConfig(config)
  }

  isConfigured(): boolean {
    return this.configured
  }

  async writeTask(
    task: CanonicalTask,
    expectedUpdatedAt: string | null,
  ): Promise<SyncWriteResult> {
    if (!this.configured) {
      throw new SupabaseTransportNotConfiguredError()
    }

    return writeSupabaseTodo(task, expectedUpdatedAt)
  }

  async pullTasks(): Promise<CanonicalTask[]> {
    if (!this.configured) {
      throw new SupabaseTransportNotConfiguredError()
    }

    return pullSupabaseTodos()
  }
}
